linera-core/src/client/validator_manager/node_info.rs

```diff
@@ -0,0 +1,189 @@
+    /// will be woken when a request completes and releases its slot.
+    pub(super) slot_available: Arc<tokio::sync::Notify>,
+
+    /// Total number of requests processed (for monitoring and cold-start handling)
+    pub(super) total_requests: u64,
+
+    /// Configuration for scoring weights
+    pub(super) weights: ScoringWeights,
+
+    /// EMA smoothing factor (0 < alpha < 1)
+    /// Higher values give more weight to recent observations
+    pub(super) alpha: f64,
+
+    /// Maximum expected latency in milliseconds for score normalization
+    pub(super) max_expected_latency_ms: f64,
+
+    /// Maximum expected in-flight requests for score normalization
+    pub(super) max_in_flight: usize,
+}
+
+impl<Env: Environment> NodeInfo<Env> {
+    /// Creates a new `NodeInfo` with optimistic initial values.
+    ///
+    /// New nodes start with:
+    /// - 100ms expected latency (reasonable default)
+    /// - 100% success rate (optimistic start)
+    /// - Zero in-flight requests
+    /// - Default weights and smoothing factor
+    /// - Default normalization bounds
+    #[allow(unused)]
+    pub(super) fn new(node: RemoteNode<Env::ValidatorNode>) -> Self {
+        Self::with_config(
+            node,
+            ScoringWeights::default(),
+            0.1,
+            MAX_ACCEPTED_LATENCY_MS,
+            MAX_IN_FLIGHT_REQUESTS,
+        )
+    }
+
+    /// Creates a new `NodeInfo` with custom configuration.
+    pub(super) fn with_config(
+        node: RemoteNode<Env::ValidatorNode>,
+        weights: ScoringWeights,
+        alpha: f64,
+        max_expected_latency_ms: f64,
+        max_in_flight: usize,
+    ) -> Self {
+        Self {
+            node,
+            ema_latency_ms: 100.0, // Start with reasonable latency expectation
+            ema_success_rate: 1.0, // Start optimistically with 100% success
+            in_flight_requests: Arc::new(tokio::sync::RwLock::new(0)),
+            slot_available: Arc::new(tokio::sync::Notify::new()),
+            total_requests: 0,
+            weights,
+            alpha: alpha.clamp(0.01, 0.5), // Ensure alpha is in reasonable range
+            max_expected_latency_ms,
+            max_in_flight,
+        }
+    }
+
+    /// Calculates a normalized performance score (0.0 to 1.0) using weighted metrics.
+    ///
+    /// The score combines three normalized components:
+    /// - **Latency score**: Inversely proportional to EMA latency
+    /// - **Success score**: Directly proportional to EMA success rate
+    /// - **Load score**: Inversely proportional to current load
+    ///
+    /// Returns a score from 0.0 to 1.0, where higher values indicate better performance.
+    pub(super) async fn calculate_score(&self) -> f64 {
+        // 1. Normalize Latency (lower is better, so we invert)
+        let latency_score = 1.0
+            - (self.ema_latency_ms.min(self.max_expected_latency_ms)
+                / self.max_expected_latency_ms);
```

Comment:
If the max latency is 5s, that means the latency scores of a node that returned after 100 ms will be very similar to that of a node that returned with 200 ms.

I think that's fine for now, especially since the requests can have very differently-sized responses anyway, but maybe something to tweak in the future.

Reply:
We can decrease the `max_expected_latency_ms` to something like 500ms and tweak the parameters to favor lower latency even more.

Reply:
Since it's now configurable I'm gonna leave it as-is. We can adjust it with different args.